### PR TITLE
Reduce the cost of loading all built-ins targets

### PR DESCRIPTION
**Pull request by Urgau**
This PR started by measuring the exact slowdown of checking of well known conditional values.
Than this PR implemented some technics to reduce the cost of loading all built-ins targets.

cf. https://github.com/rust-lang/rust/issues/82450#issuecomment-1073992323

### CODE REVIEW

**Comment by petrochenkov** [comment]
@Urgau 
Yes, non-empty `String`s are the main blocker for constification of target specs, but there may be other issues.

**Comment by Urgau** [comment]
Before measuring the cost (hopefully negative) of replacing every `String` by `Cow<String>`, I would be good to check whenever it's actually significative. I just push a commit who disable the checking but NOT the loading of values (targets, ...).

**Comment by compiler-errors** [comment]
Er, actually it has failing tests. Probably will fail bors-try.

**Comment by Urgau** [comment]
Perf regressions are decreasing (-0.4% on average) but unfortunately they are still quite important.
Let's try to also replace the `Vec` by their `Cow` equivalent.

**Comment by Urgau** [comment]
Perf regressions slightly are decreasing (-0.1% on average but largest gone from +2.7% to +2.3%) but still quite important.
Let's try to also replace CrtObjects (alias for a BTreeMap) with a custom struct with Cow.

**Comment by Urgau** [comment]
Oh! This was expected, this a regression +0.2% (+1% to 1.2%) from before.
I was gonna try to also create a custom struct for `LinkArgs` but instead let's just confirm that the regression is indeed due to the custom struct (maybe it's because it has became much bigger ?).

**Comment by Urgau** [comment]
@petrochenkov I think I'm at the limit of what can be accomplish without having const-allocation, const Vec, const BTreeMap, const Default trait, ...
If you have other ideas that you want me to try out tell me ? otherwise would you want me to turn this PR into a actual merge-able PR ?

**Comment by petrochenkov** [comment]
So this PR doesn't (partially) build the target spec list at compile time yet, but rather makes it faster to build at runtime by avoiding a bunch of allocations by using `Cow`s.
This seems fine by itself, the penalty from enabling cfg value checking for hello-world is almost halved.
`Cow`s instead of `String`s and `Vec`tors also do not make the code noticeably more complex, so it's mostly a win-win situation.

>const Default trait

Unlike const allocations this doesn't seem to be the problem though.

**Comment by petrochenkov** [comment]
I think 2% regression on hello-world would be an acceptable regression for enabling this useful functionality, especially given that further improvements can be made (building a part of the list at compile time, and almost the whole list some time later when const allocations arrive).

**Comment by petrochenkov** [comment]
What percent of `fn target` functions can be converted to `const fn` now?
(Assuming that `Default::default` is converted to a const impl.)

**Comment by Urgau** [comment]
> What percent of `fn target` functions can be converted to `const fn` now? (Assuming that `Default::default` is converted to a const impl.)

@petrochenkov  None ! They all depend at-least on const `Default` (for `..Default::default`), const `Into` (for `.into()` used to get the `Cow`, can be replaced by `Cow::Borrowed`), and most importantly const destructor (for the `..` syntax).

<details>
<summary> Errors when trying to constify armv7-unknown-linux-gnueabi </summary>

```text
error[E0015]: cannot call non-const fn `<&str as Into<Cow<str>>>::into` in constant functions
 --> compiler/rustc_target/src/spec/linux_base.rs:7:21
  |
7 |         os: "linux".into(),
  |                     ^^^^^^
  |
  = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants

error[E0015]: cannot call non-const fn `<TargetOptions as Default>::default` in constant functions
  --> compiler/rustc_target/src/spec/linux_base.rs:16:11
   |
16 |         ..Default::default()
   |           ^^^^^^^^^^^^^^^^^^
   |
   = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants

error[E0493]: destructors cannot be evaluated at compile-time
  --> compiler/rustc_target/src/spec/linux_base.rs:16:11
   |
16 |         ..Default::default()
   |           ^^^^^^^^^^^^^^^^^^ constant functions cannot evaluate destructors
17 |     }
18 | }
   | - value is dropped here

error[E0015]: cannot call non-const fn `<&str as Into<Cow<str>>>::into` in constant functions
 --> compiler/rustc_target/src/spec/linux_gnu_base.rs:4:32
  |
4 |     TargetOptions { env: "gnu".into(), ..super::linux_base::opts() }
  |                                ^^^^^^
  |
  = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants

error[E0493]: destructors cannot be evaluated at compile-time
 --> compiler/rustc_target/src/spec/linux_gnu_base.rs:4:42
  |
4 |     TargetOptions { env: "gnu".into(), ..super::linux_base::opts() }
  |                                          ^^^^^^^^^^^^^^^^^^^^^^^^^ constant functions cannot evaluate destructors
5 | }
  | - value is dropped here

error[E0015]: cannot call non-const fn `<&str as Into<Cow<str>>>::into` in constant functions
 --> compiler/rustc_target/src/spec/armv7_unknown_linux_gnueabihf.rs:8:54
  |
8 |         llvm_target: "armv7-unknown-linux-gnueabihf".into(),
  |                                                      ^^^^^^
  |
  = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants

error[E0015]: cannot call non-const fn `<&str as Into<Cow<str>>>::into` in constant functions
  --> compiler/rustc_target/src/spec/armv7_unknown_linux_gnueabihf.rs:10:76
   |
10 |         data_layout: "e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64".into(),
   |                                                                            ^^^^^^
   |
   = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants

error[E0015]: cannot call non-const fn `<&str as Into<Cow<str>>>::into` in constant functions
  --> compiler/rustc_target/src/spec/armv7_unknown_linux_gnueabihf.rs:11:21
   |
11 |         arch: "arm".into(),
   |                     ^^^^^^
   |
   = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants

error[E0015]: cannot call non-const fn `<&str as Into<Cow<str>>>::into` in constant functions
  --> compiler/rustc_target/src/spec/armv7_unknown_linux_gnueabihf.rs:13:27
   |
13 |             abi: "eabihf".into(),
   |                           ^^^^^^
   |
   = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants

error[E0015]: cannot call non-const fn `<&str as Into<Cow<str>>>::into` in constant functions
  --> compiler/rustc_target/src/spec/armv7_unknown_linux_gnueabihf.rs:15:54
   |
15 |             features: "+v7,+vfp3,-d32,+thumb2,-neon".into(),
   |                                                      ^^^^^^
   |
   = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants

error[E0015]: cannot call non-const fn `<&str as Into<Cow<str>>>::into` in constant functions
  --> compiler/rustc_target/src/spec/armv7_unknown_linux_gnueabihf.rs:17:44
   |
17 |             mcount: "\u{1}__gnu_mcount_nc".into(),
   |                                            ^^^^^^
   |
   = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants

error[E0493]: destructors cannot be evaluated at compile-time
  --> compiler/rustc_target/src/spec/armv7_unknown_linux_gnueabihf.rs:18:15
   |
18 |             ..super::linux_gnu_base::opts()
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ constant functions cannot evaluate destructors
...
21 | }
   | - value is dropped here

Some errors have detailed explanations: E0015, E0493.
For more information about an error, try `rustc --explain E0015`.
error: could not compile `rustc_target` due to 12 previous errors
```

</details>

File: compiler/rustc_target/src/spec/mod.rs [hunk 1]
```diff
@@ -1026,6 +1027,25 @@ supported_targets! {
     ("mips64-openwrt-linux-musl", mips64_openwrt_linux_musl),
 }
 
+/// Cow-Vec-Str: Cow<'static, [Cow<'static, str>]>
+// FIXME(Urgau): Figure out why the obvious form `["".into()].into()` doesn't work.
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/mod.rs:1031]: This FIXME is not very actionable, probably better to remove it.

File: compiler/rustc_target/src/spec/mod.rs [hunk 2]
```diff
@@ -1026,6 +1027,25 @@ supported_targets! {
...
+macro_rules! cvs {
+    () => {
+        ::std::borrow::Cow::Borrowed(&[])
+    };
+    ($($x:expr),+ $(,)?) => {
+        {
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/mod.rs:1037]: This extra `{}` block looks unnecessary.

File: compiler/rustc_target/src/spec/mod.rs [hunk 3]
```diff
@@ -1064,14 +1084,14 @@ impl TargetWarnings {
 #[derive(PartialEq, Clone, Debug)]
 pub struct Target {
     /// Target triple to pass to LLVM.
-    pub llvm_target: String,
+    pub llvm_target: Cow<'static, str>,
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/mod.rs:1087]: `Cow<'static, T>` is used so much in this file that it make sense to introduce a private alias `type StaticCow<T> = Cow<'static, T>`.

File: compiler/rustc_target/src/spec/mod.rs [hunk 4]
```diff
@@ -1437,20 +1457,20 @@ impl Default for TargetOptions {
...
             is_builtin: false,
             endian: Endian::Little,
-            c_int_width: "32".to_string(),
-            os: "none".to_string(),
-            env: String::new(),
-            abi: String::new(),
-            vendor: "unknown".to_string(),
+            c_int_width: "32".into(),
+            os: "none".into(),
+            env: Cow::from(""),
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/mod.rs:1462]:
```suggestion
            env: "".into(),
```
Style nit: the `"string".into()` seems more common for constructing cows, so maybe it's better to normalize to it.

File: compiler/rustc_target/src/spec/mod.rs [hunk 5]
```diff
@@ -1437,20 +1457,20 @@ impl Default for TargetOptions {
...
-            linker: option_env!("CFG_DEFAULT_LINKER").map(|s| s.to_string()),
+            linker: option_env!("CFG_DEFAULT_LINKER").map(|s| s.into()),
             lld_flavor: LldFlavor::Ld,
             pre_link_args: LinkArgs::new(),
             post_link_args: LinkArgs::new(),
             link_script: None,
-            asm_args: Vec::new(),
-            cpu: "generic".to_string(),
-            features: String::new(),
+            asm_args: Cow::Borrowed(&[]),
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/mod.rs:1471]:
```suggestion
            asm_args: cvs![],
```
?

(Here and in a few other places.)

File: compiler/rustc_target/src/spec/x86_64_fortanix_unknown_sgx.rs [hunk 6]
```diff
@@ -1,4 +1,6 @@
-use std::iter;
+use std::{borrow::Cow, iter};
+
+use crate::spec::cvs;
 
 use super::{LinkerFlavor, LldFlavor, Target, TargetOptions};
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/x86_64_fortanix_unknown_sgx.rs:5]:
```suggestion
use super::{cvs, LinkerFlavor, LldFlavor, Target, TargetOptions};
```
Style: all `cvs` imports can be merged into already existing imports.

File: compiler/rustc_target/src/spec/apple_base.rs [hunk 7]
```diff
@@ -51,7 +53,7 @@ pub fn opts(os: &str) -> TargetOptions {
         // this environment variable too in recent versions.
         //
         // For some more info see the commentary on #47086
-        link_env: vec![("ZERO_AR_DATE".to_string(), "1".to_string())],
+        link_env: Cow::Borrowed(&[(Cow::Borrowed("ZERO_AR_DATE"), Cow::Borrowed("1"))]),
```
Comment (petrochenkov) [inline compiler/rustc_target/src/spec/apple_base.rs:56]:
```suggestion
        link_env: cfs!["ZERO_AR_DATE", "1"],
```

File: compiler/rustc_session/src/config.rs [hunk 8]
```diff
@@ -1037,7 +1037,7 @@ impl<T> Default for CheckCfg<T> {
         CheckCfg {
             names_valid: Default::default(),
             values_valid: Default::default(),
-            well_known_values: false,
+            well_known_values: true,
```
Comment (petrochenkov) [inline compiler/rustc_session/src/config.rs:1040]: This needs to be reverted.

File: compiler/rustc_session/src/config.rs [hunk 9]
```diff
@@ -956,7 +956,7 @@ fn default_configuration(sess: &Session) -> CrateConfig {
     ret.reserve(7); // the minimum number of insertions
     // Target bindings.
     ret.insert((sym::target_os, Some(Symbol::intern(os))));
-    for fam in &sess.target.families {
+    for fam in sess.target.families.iter() {
```
Comment (petrochenkov) [inline compiler/rustc_session/src/config.rs:959]: Unnecessary change?

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 10]
```diff
@@ -1216,7 +1217,7 @@ pub fn linker_and_flavor(sess: &Session) -> (PathBuf, LinkerFlavor) {
 
     if let Some(ret) = infer_from(
         sess,
-        sess.target.linker.clone().map(PathBuf::from),
+        sess.target.linker.as_ref().map(|l| PathBuf::from(l.as_ref())),
```
Comment (petrochenkov) [inline compiler/rustc_codegen_ssa/src/back/link.rs:1220]: Are 2 `as_refs` necessary here?

**Comment by petrochenkov** [comment]
The PR's title and description need to be updated.
Also CI is failing.
@bors rollup=maybe

File: compiler/rustc_target/src/spec/apple_base.rs (hunk 7 above)
Comment (Urgau) [reply compiler/rustc_target/src/spec/apple_base.rs:54]: This won't work as the `cvs!` macro is for Cow-Vec-Str not Cow-Vec-Tuple-Str.

File: compiler/rustc_session/src/config.rs (hunk 9 above)
Comment (Urgau) [reply compiler/rustc_session/src/config.rs]: Unfortunately, no because:
```text
error[E0277]: `&Cow<'_, [Cow<'static, str>]>` is not an iterator
   --> compiler/rustc_session/src/config.rs:959:16
    |
959 |     for fam in &sess.target.families {
    |                ^^^^^^^^^^^^^^^^^^^^^ `&Cow<'_, [Cow<'static, str>]>` is not an iterator
    |
    = help: the trait `Iterator` is not implemented for `&Cow<'_, [Cow<'static, str>]>`
    = note: required because of the requirements on the impl of `IntoIterator` for `&Cow<'_, [Cow<'static, str>]>`
```
But still changed to `as_ref()` instead of `iter()`.

File: compiler/rustc_codegen_ssa/src/back/link.rs (hunk 10 above)
Comment (Urgau) [reply compiler/rustc_codegen_ssa/src/back/link.rs]: Seems like yes and no, the first one is responsible for `Option<Cow<'static, str>>` to `Option<&Cow<'static, str>>` and the second one for `&Cow<'static, str>` to `&str>` but it can be replaced by `as_deref()` with does a similar thing: `Option<Cow<'static, str>>` to `&Option<&str>`.

**Comment by petrochenkov** [comment]
Unit tests in `rustc_target` are failing.
@rustbot author

File: compiler/rustc_target/src/spec/mod.rs [hunk 11]
```diff
@@ -1107,6 +1104,8 @@ impl HasTargetSpec for Target {
     }
 }
 
+type StaticCow<T> = Cow<'static, T>;
```
Comment (bjorn3) [inline compiler/rustc_target/src/spec/mod.rs:1107]: I don't think this adds much. It only saves 2 characters.

**Comment by Urgau** [comment]
Review comments addressed or responded. Commits squashed to only 4, but can squash to only 1 if you prefer.
Title and PR description also addressed. Also fix all of the failing unit tests (missing `to_string()`).

@rustbot ready

File: compiler/rustc_target/src/spec/mod.rs (hunk 11 above)
Comment (petrochenkov) [reply compiler/rustc_target/src/spec/mod.rs:1107]: I asked to add it in https://github.com/rust-lang/rust/pull/95202#discussion_r841230397 because it looks less noisy due to removing punctuation like `'` and `,` even if it's only slightly shorter. (Not a strong opinion.)
